### PR TITLE
fix: report exact ring cardinality and cover range boundaries

### PR DESCRIPTION
**Pull request by ReubenBond**
## Motivation

Grain directory partitioning depends on accurate range boundaries and point counts. A 32-bit ring contains 2^32 points, so both full-ring cardinality and almost-full ranges need distinct, exact sizes. Exact counts preserve the size-conservation invariant used during membership updates, including transitions between full and partial ownership.

This PR corrects the sizing contract and adds unit, CsCheck property-based, and boundary-value coverage for `RingRange` and `RingRangeCollection`.

## Summary of Changes

- Return exact `ulong` point counts from both `Size` properties, including 2^32 for a full ring. Include zero when counting wrapped ranges and sum non-overlapping collection ranges in one pass.
- Derive collection fullness from exact cardinality, calculate percentages against the full-ring count, and reuse `RingRange.Size` for directory query batching. Range endpoints remain `uint` and serialized fields retain their existing IDs.
- Exercise construction, single-point ranges, containment, comparison, intersections, complements, differences, equality, hashing, enumeration, and formatting using unit tests, CsCheck properties, and boundary values.
- Compare generated sizes with an independent inclusive-interval model. Cover full partitions, single-point gaps, and boundary cardinalities; assert exactly 2^32 covered points in directory membership tests.
- Give equal empty collection representations matching hash codes and make default collections enumerate as empty through both the struct enumerator and interface-based enumeration. Add regressions for both behaviors.

### CODE REVIEW

File: src/Orleans.Runtime/GrainDirectory/RingRangeCollection.cs [hunk 1]
```diff
@@ -62,9 +62,9 @@ public static RingRangeCollection Create<TCollection>(TCollection ranges) where
 
     public bool IsEmpty => Ranges.Length == 0 || Ranges.All(r => r.IsEmpty);
 
-    public bool IsFull => !IsEmpty && Ranges.Sum(r => r.Size) == uint.MaxValue;
+    public bool IsFull => !IsEmpty && Ranges.Sum(static r => (long)r.Size) >= uint.MaxValue;
 
-    public uint Size => (uint)Ranges.Sum(static r => r.Size);
+    public uint Size => Ranges.Sum(static r => (long)r.Size) >= uint.MaxValue ? uint.MaxValue : (uint)Ranges.Sum(static r => (long)r.Size);
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/GrainDirectory/RingRangeCollection.cs]: Refactored `Size` to sum ranges using a single `foreach` loop over `Ranges` (using `ImmutableArray<RingRange>`'s struct enumerator to avoid LINQ/lambda allocations) and updated `IsFull` to check `Size == uint.MaxValue && !IsEmpty`.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/GrainDirectory/RingRangeCollection.cs [hunk 2]
```diff
@@ -62,9 +62,26 @@ public static RingRangeCollection Create<TCollection>(TCollection ranges) where
 
     public bool IsEmpty => Ranges.Length == 0 || Ranges.All(r => r.IsEmpty);
 
-    public bool IsFull => !IsEmpty && Ranges.Sum(r => r.Size) == uint.MaxValue;
+    public bool IsFull => !IsEmpty && Size == uint.MaxValue;
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/GrainDirectory/RingRangeCollection.cs]: Updated `RingRangeCollection.IsFull` to determine fullness based on range continuity rather than checking `Size == uint.MaxValue`.

For a single range, it returns `Ranges[0].IsFull` (so an almost-full range like `(1, 0]` with `Size == uint.MaxValue` now correctly evaluates `IsFull` to `false`). For multiple ranges, it verifies that adjacent ranges are contiguous (`Ranges[i].End == Ranges[i + 1].Start`) and that the last range wraps around to meet the first (`Ranges[^1].End == Ranges[0].Start`).

Also added unit tests for `(1, 0]`, contiguous multi-range full collections, and multi-range collections with gaps.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: test/Orleans.Core.Tests/Directory/RingRangeCollectionTests.cs [hunk 3]
```diff
@@ -231,5 +236,233 @@ public void Formatting_ReportsSubrangeCountAndSize()
...
+                long expectedSum = 0;
+                foreach (var r in collection.Ranges)
+                {
+                    expectedSum += r.Size;
+                }
+
+                uint expectedSize = expectedSum >= uint.MaxValue ? uint.MaxValue : (uint)expectedSum;
+                Assert.Equal(expectedSize, collection.Size);
+                Assert.Equal(expectedSize == uint.MaxValue, collection.IsFull);
+            }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Orleans.Core.Tests/Directory/RingRangeCollectionTests.cs]: Fixed in d24578838d. Both `Size` properties now return exact `ulong` point counts, with `1UL << 32` representing a full ring. `Property_SizeAndIsFull` computes its expected count using an independent inclusive-interval model and checks fullness against that exact full-ring count.

Boundary cases cover `(0, uint.MaxValue]` and `(1, 0]` as almost-full ranges, while complementary ranges and generated complete partitions total exactly 2^32. The directory membership properties also assert exact full-ring totals.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).
